File: frontend/src/types/workTypeReport.ts
/** TS types for thematic work-type report — mirrors app/schemas/work_type_report.py */

export type GroupingDim = 'theme' | 'team' | 'role' | 'employee' | 'project' | 'issue';

// ---- Themes ----

export interface ThemeOut {
  id: string;
  work_type_id: string;
  name: string;
  description: string | null;
  color: string;
  sort_order: number;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface ThemeCreateRequest {
  work_type_id: string;
  name: string;
  description?: string | null;
  color?: string;
  sort_order?: number;
}

export interface ThemeUpdateRequest {
  name?: string;
  description?: string | null;
  color?: string;
  sort_order?: number;
}

export interface ThemeMergeRequest {
  target_theme_id: string;
}

/** Кандидат из ведра «Другое» — возвращается /themes GET вместе со словарём */
export interface ThemeCandidate {
  proposed_name: string;
  issues_count: number;
  hours: number;
  sample_keys: string[];
  snapshot_id: string;
}

export interface ThemeListResponse {
  themes: ThemeOut[];
  candidates: ThemeCandidate[];
}

// ---- Snapshot data shapes ----

export interface ThemeEmployeeBreakdown {
  employee_id: string;
  hours: number;
  name: string;
  role: string;
  team: string;
}

export interface ThemeTopTask {
  key: string;
  summary: string;
  hours: number;
  contribution: string | null;
}

export interface ThemeIssue {
  issue_id: string;
  key: string;
  summary: string;
  hours: number;
  contribution: string | null;
  employee_breakdown: { name: string; role: string; team: string; hours: number }[];
}

export interface ThemeTotals {
  hours: number;
  pct: number;
  tasks_count: number;
  employees_count: number;
}

export interface Theme {
  theme_id: string | null;
  name: string;
  color: string;
  is_new: boolean;
  is_low_confidence?: boolean;
  totals: ThemeTotals;
  by_employee: ThemeEmployeeBreakdown[];
  top_tasks: ThemeTopTask[];
  issues: ThemeIssue[];
  evidence_keys: string[];
  narrative: string;
}

export interface Outlier {
  key: string;
  issue_id: string;
  reason: string;
  value: number | string;
  context: Record<string, unknown> | null;
  explanation: string;
}

/** Кандидат во frontend-вариации (без snapshot_id — он живёт на корне ответа) */
export interface Candidate {
  proposed_name: string;
  issues_count: number;
  hours: number;
  sample_keys: string[];
}

export interface ManualReviewIssue {
  issue_id: string;
  key: string;
  summary: string;
  hours: number;
  failure_reason: string;
}

export interface ReportTotals {
  hours: number;
  tasks: number;
  employees: number;
  themes_count: number;
}

export interface WorkTypeReportData {
  headline: string;
  totals: ReportTotals;
  themes: Theme[];
  candidates: Candidate[];
  outliers: Outlier[];
  recommendation: { text: string; expected_impact: string };
  manual_review_required: ManualReviewIssue[];
  is_fallback_narrative: boolean;
}

// ---- Report request / response ----

export interface WorkTypeReportRequest {
  work_type_id: string;
  year: number;
  quarter: number;
  month?: number | null;
  teams?: string[];
  force_refresh?: boolean;
}

export interface WorkTypeReportResponse {
  snapshot_id: string;
  work_type_id: string;
  year: number;
  quarter: number;
  month: number | null;
  start_date: string;
  end_date: string;
  team_set: string[];
  generated_at: string;
  model_id: string | null;
  prompt_version: string | null;
  dictionary_version: number;
  is_stale: boolean;
  data: WorkTypeReportData;
}

// ---- Candidate actions ----

export interface CandidateAcceptRequest {
  snapshot_id: string;
  proposed_name: string;
  new_theme_name?: string | null;
  color?: string;
}

export interface CandidateMergeRequest {
  snapshot_id: string;
  proposed_name: string;
  target_theme_id: string;
}

export interface CandidateIgnoreRequest {
  snapshot_id: string;
  proposed_name: string;
}

export interface ManualClassifyRequest {
  issue_id: string;
  work_type_id: string;
  theme_id?: string | null;
  contribution_text?: string | null;
}

// ---- Layouts ----

export interface LayoutOut {
  id: string;
  user_id: string;
  work_type_id: string;
  name: string;
  grouping_dims: GroupingDim[];
  visible_columns: string[] | null;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface LayoutCreateRequest {
  work_type_id: string;
  name: string;
  grouping_dims: GroupingDim[];
  visible_columns?: string[] | null;
  is_default?: boolean;
}

export interface LayoutUpdateRequest {
  name?: string;
  grouping_dims?: GroupingDim[];
  visible_columns?: string[] | null;
  is_default?: boolean;
}

File: app/schemas/work_type_report.py
"""Pydantic schemas for thematic work-type report."""
from datetime import date, datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


# ---- Themes ----

class ThemeBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    color: str = Field(default="#00c9c8", pattern=r"^#[0-9A-Fa-f]{6}$")
    sort_order: int = 0


class ThemeCreateRequest(ThemeBase):
    work_type_id: str


class ThemeUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    color: Optional[str] = Field(None, pattern=r"^#[0-9A-Fa-f]{6}$")
    sort_order: Optional[int] = None


class ThemeMergeRequest(BaseModel):
    target_theme_id: str


class ThemeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    work_type_id: str
    name: str
    description: Optional[str]
    color: str
    sort_order: int
    is_archived: bool
    aliases: list[str] = []
    created_at: datetime
    updated_at: datetime


class AliasAddRequest(BaseModel):
    alias: str = Field(..., min_length=1, max_length=255)


class ThemeAliasResponse(BaseModel):
    theme_id: str
    aliases: list[str]


class ThresholdResponse(BaseModel):
    threshold: float


class ThresholdRequest(BaseModel):
    threshold: float = Field(..., ge=0.0, le=1.0)


class ThemeCandidate(BaseModel):
    """Кандидат в словарь — из ведра «Другое» свежего снапшота."""
    proposed_name: str
    issues_count: int
    hours: float
    sample_keys: list[str]
    snapshot_id: str


class ThemeListResponse(BaseModel):
    themes: list[ThemeOut]
    candidates: list[ThemeCandidate]


# ---- Report ----

class WorkTypeReportRequest(BaseModel):
    work_type_id: str
    year: int = Field(..., ge=2020, le=2100)
    quarter: int = Field(..., ge=1, le=4)
    month: Optional[int] = Field(None, ge=1, le=12)
    teams: list[str] = []
    force_refresh: bool = False


class WorkTypeReportResponse(BaseModel):
    snapshot_id: str
    work_type_id: str
    year: int
    quarter: int
    month: Optional[int]
    start_date: date
    end_date: date
    team_set: list[str]
    generated_at: datetime
    model_id: Optional[str]
    prompt_version: Optional[str]
    dictionary_version: int
    is_stale: bool
    data: dict[str, Any]


class CandidateAcceptRequest(BaseModel):
    snapshot_id: str
    proposed_name: str
    new_theme_name: Optional[str] = None
    color: str = Field(default="#00c9c8", pattern=r"^#[0-9A-Fa-f]{6}$")


class CandidateMergeRequest(BaseModel):
    snapshot_id: str
    proposed_name: str
    target_theme_id: str


class CandidateIgnoreRequest(BaseModel):
    snapshot_id: str
    proposed_name: str


class ManualClassifyRequest(BaseModel):
    issue_id: str
    work_type_id: str
    theme_id: Optional[str]   # None = leave unclassified
    contribution_text: Optional[str] = None


# ---- Layouts ----

class LayoutBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    grouping_dims: list[Literal["theme", "team", "role", "employee", "project", "issue"]]
    visible_columns: Optional[list[str]] = None
    is_default: bool = False


class LayoutCreateRequest(LayoutBase):
    work_type_id: str


class LayoutUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    grouping_dims: Optional[list[Literal["theme", "team", "role", "employee", "project", "issue"]]] = None
    visible_columns: Optional[list[str]] = None
    is_default: Optional[bool] = None


class LayoutOut(LayoutBase):
    model_config = ConfigDict(from_attributes=True)
    id: str
    user_id: str
    work_type_id: str
    created_at: datetime
    updated_at: datetime
